contrib/juliac.jl

```diff
@@ -55,55 +54,77 @@ isnothing(outname) && error("No output file specified")
+    verbose && println("Running: $cmd")
+    if !success(pipeline(cmd; stdout, stderr))
+        println(stderr, "\nFailed to compile $file")
+        exit(1)
+    end
 
-if output_type == "--output-lib" || output_type == "--output-sysimage"
-    of, ext = splitext(outname)
-    soext = "." * Base.BinaryPlatforms.platform_dlext()
-    if ext == ""
-        outname = of * soext
+    # Compile the initialization code
+    open(initsrc_path, "w") do io
+        print(io, """
+                  #include <julia.h>
+                  __attribute__((constructor)) void static_init(void) {
+                      if (jl_is_initialized())
+                          return;
+                      julia_init(JL_IMAGE_IN_MEMORY);
+                      jl_exception_clear();
+                  }
+                  """)
     end
+    run(`cc $(cflags) -g -c -o $init_path $initsrc_path`)
 end
 
-julia_libs = Base.shell_split(Base.isdebugbuild() ? "-ljulia-debug -ljulia-internal-debug" : "-ljulia -ljulia-internal")
-try
-    if output_type == "--output-lib"
-        run(`cc $(allflags) -o $outname -shared -Wl,$(Base.Linking.WHOLE_ARCHIVE) $img_path  -Wl,$(Base.Linking.NO_WHOLE_ARCHIVE) $init_path  $(julia_libs)`)
-    elseif output_type == "--output-sysimage"
-        run(`cc $(allflags) -o $outname -shared -Wl,$(Base.Linking.WHOLE_ARCHIVE) $img_path  -Wl,$(Base.Linking.NO_WHOLE_ARCHIVE)             $(julia_libs)`)
-    else
-        run(`cc $(allflags) -o $outname -Wl,$(Base.Linking.WHOLE_ARCHIVE) $img_path -Wl,$(Base.Linking.NO_WHOLE_ARCHIVE) $init_path $(julia_libs)`)
+function link_products()
+    global outname
+    if output_type == "--output-lib" || output_type == "--output-sysimage"
+        of, ext = splitext(outname)
+        soext = "." * Base.BinaryPlatforms.platform_dlext()
+        if ext == ""
+            outname = of * soext
+        end
+    end
+
+    julia_libs = Base.shell_split(Base.isdebugbuild() ? "-ljulia-debug -ljulia-internal-debug" : "-ljulia -ljulia-internal")
+    try
+        if output_type == "--output-lib"
+            run(`cc $(allflags) -o $outname -shared -Wl,$(Base.Linking.WHOLE_ARCHIVE) $img_path  -Wl,$(Base.Linking.NO_WHOLE_ARCHIVE) $init_path  $(julia_libs)`)
+        elseif output_type == "--output-sysimage"
+            run(`cc $(allflags) -o $outname -shared -Wl,$(Base.Linking.WHOLE_ARCHIVE) $img_path  -Wl,$(Base.Linking.NO_WHOLE_ARCHIVE)             $(julia_libs)`)
+        else
+            run(`cc $(allflags) -o $outname -Wl,$(Base.Linking.WHOLE_ARCHIVE) $img_path -Wl,$(Base.Linking.NO_WHOLE_ARCHIVE) $init_path $(julia_libs)`)
+        end
+    catch e
+        println("\nCompilation failed: ", e)
+        exit(1)
     end
-catch
-    println("\nCompilation failed.")
-    exit(1)
 end
+
+precompile_env()
```

Comment:
maybe a main function 😄

Reply:
sure, but a separate change maybe - it requires re-factoring all the global shenanigans here

Reply:
Fair enough